File: backend/app/schemas/admin/campaigns.py
"""Campaign management schemas for Phase 5."""

from datetime import date, datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ============================================================================
# Campaign Schemas
# ============================================================================

class CampaignBase(BaseModel):
    """Base campaign schema with common fields."""
    
    name: str = Field(..., min_length=1, max_length=255, description="Campaign name")
    description: Optional[str] = Field(None, description="Campaign description")
    message_template: str = Field(..., min_length=1, description="Message template for TCA")
    trigger_rules: Optional[Dict[str, Any]] = Field(None, description="JSON trigger configuration")
    target_audience: Optional[Dict[str, Any]] = Field(None, description="JSON audience targeting")
    status: str = Field("draft", description="Campaign status (draft, active, paused, completed)")
    priority: str = Field("medium", description="Campaign priority (low, medium, high)")


class CampaignCreate(CampaignBase):
    """Schema for creating a new campaign."""
    pass


class CampaignUpdate(BaseModel):
    """Schema for updating a campaign."""
    
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    message_template: Optional[str] = Field(None, min_length=1)
    trigger_rules: Optional[Dict[str, Any]] = None
    target_audience: Optional[Dict[str, Any]] = None
    status: Optional[str] = None
    priority: Optional[str] = None


class CampaignResponse(CampaignBase):
    """Schema for campaign response."""
    
    id: UUID
    created_by: Optional[int]
    created_at: datetime
    updated_at: datetime
    last_executed_at: Optional[datetime]
    
    model_config = ConfigDict(from_attributes=True)


class CampaignListResponse(BaseModel):
    """Schema for paginated campaign list."""
    
    items: List[CampaignResponse]
    total: int
    skip: int
    limit: int


class CampaignSummary(BaseModel):
    """Schema for campaign statistics summary."""
    
    total: int
    draft: int
    active: int
    paused: int
    completed: int


# ============================================================================
# Campaign Trigger Schemas
# ============================================================================

class CampaignTriggerBase(BaseModel):
    """Base trigger schema."""
    
    condition_type: str = Field(..., description="Condition type (ia_insight, manual, scheduled, threshold_breach)")
    condition_value: Dict[str, Any] = Field(..., description="JSON condition definition")
    evaluation_frequency: str = Field("daily", description="Evaluation frequency (hourly, daily, weekly)")


class CampaignTriggerCreate(CampaignTriggerBase):
    """Schema for creating a trigger."""
    
    campaign_id: UUID


class CampaignTriggerResponse(CampaignTriggerBase):
    """Schema for trigger response."""
    
    id: UUID
    campaign_id: UUID
    last_evaluated_at: Optional[datetime]
    last_match_at: Optional[datetime]
    match_count: int
    
    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Campaign Metrics Schemas
# ============================================================================

class CampaignMetricsResponse(BaseModel):
    """Schema for campaign metrics response."""
    
    id: UUID
    campaign_id: UUID
    execution_date: date
    messages_sent: int
    users_targeted: int
    users_engaged: int
    success_rate: Optional[float]
    avg_sentiment_before: Optional[float]
    avg_sentiment_after: Optional[float]
    
    model_config = ConfigDict(from_attributes=True)


class CampaignMetricsList(BaseModel):
    """Schema for campaign metrics list."""
    
    campaign_id: UUID
    metrics: List[CampaignMetricsResponse]
    total_messages_sent: int
    total_users_targeted: int
    total_users_engaged: int
    average_success_rate: Optional[float]


# ============================================================================
# Campaign Execution Schemas
# ============================================================================

class CampaignExecutionRequest(BaseModel):
    """Schema for campaign execution request."""
    
    campaign_id: UUID
    dry_run: bool = Field(False, description="If true, only preview targets without sending")


class CampaignExecutionResponse(BaseModel):
    """Schema for campaign execution response."""
    
    success: bool
    dry_run: bool = False
    targets_count: int
    messages_sent: int
    messages_failed: int = 0
    users_targeted: int = 0
    target_user_ids: List[int] = Field(default_factory=list)
    execution_time_seconds: float = Field(0.0, description="Execution duration in seconds")
    error: Optional[str] = None


class TargetAudiencePreviewRequest(BaseModel):
    """Schema for target audience preview request."""
    
    target_criteria: Optional[Dict[str, Any]]
    limit: int = Field(50, ge=1, le=500)


class TargetAudiencePreviewResponse(BaseModel):
    """Schema for target audience preview response."""
    
    total_targets: int
    preview_limit: int
    preview: List[Dict[str, Any]]


class UserEngagementRecord(BaseModel):
    """Schema for recording user engagement."""
    
    campaign_id: UUID
    user_id: int
    engaged: bool = True


# ============================================================================
# Campaign History/Execution Log Schemas
# ============================================================================

class CampaignExecutionHistoryResponse(BaseModel):
    """Schema for campaign execution history record."""
    
    id: UUID
    campaign_id: UUID
    campaign_name: str
    executed_at: datetime
    executed_by: Optional[int]
    total_targeted: int
    messages_sent: int
    messages_failed: int
    execution_time_seconds: float
    dry_run: bool
    targeted_user_ids: Optional[List[int]] = None
    message_content: str
    error_message: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)


class CampaignExecutionHistoryListResponse(BaseModel):
    """Schema for paginated campaign execution history."""
    
    items: List[CampaignExecutionHistoryResponse]
    total: int
    skip: int
    limit: int

File: frontend/src/types/admin/campaigns.ts
/**
 * Campaign Management Types
 * Type definitions for proactive outreach campaign system
 */

export type CampaignStatus = 'draft' | 'active' | 'paused' | 'completed' | 'cancelled';

export type CampaignPriority = 'low' | 'medium' | 'high';

export type TargetAudience = 'all_users' | 'high_risk' | 'inactive' | 'recent_cases' | 'custom';

/**
 * Backend target_audience structure
 * Backend stores target_audience as a Dict/object with at minimum a 'type' field
 */
export interface TargetAudienceObject {
  type: TargetAudience;
  [key: string]: unknown; // Allow additional fields for future extensibility
}

export type TriggerType = 
  | 'sentiment_threshold' 
  | 'case_count' 
  | 'sla_breach' 
  | 'risk_score' 
  | 'inactivity';

export type ConditionOperator = 
  | 'equals' 
  | 'not_equals' 
  | 'greater_than' 
  | 'greater_than_or_equal' 
  | 'less_than' 
  | 'less_than_or_equal';

export interface TriggerConditions {
  operator?: ConditionOperator;
  value?: number;
  time_period_days?: number;
  min_breaches?: number;
  severity?: string;
  days_inactive?: number;
  had_previous_activity?: boolean;
}

export interface CampaignTrigger {
  id?: number;
  campaign_id?: string;
  trigger_type: TriggerType;
  conditions: TriggerConditions;
  is_active?: boolean;
  created_at?: string;
}

export interface CampaignMetric {
  id: string;
  campaign_id: string;
  execution_date: string;
  messages_sent: number;
  users_targeted: number;
  users_engaged: number;
  success_rate?: number;
  avg_sentiment_before?: number;
  avg_sentiment_after?: number;
}

/**
 * Campaign response from backend
 * Backend returns target_audience as an object (TargetAudienceObject)
 * or optionally as a simple Dict with unknown structure
 */
export interface Campaign {
  id: string;
  name: string;
  description?: string;
  target_audience: TargetAudienceObject | Record<string, unknown> | null;
  message_template: string;
  status: CampaignStatus;
  priority?: CampaignPriority;
  schedule?: string;
  created_at: string;
  updated_at: string;
  created_by?: number;
  last_executed_at?: string;
  triggers?: CampaignTrigger[];
}

export interface CampaignListResponse {
  items: Campaign[];
  total: number;
  skip: number;
  limit: number;
}

export interface CreateCampaignRequest {
  name: string;
  description?: string;
  target_audience: TargetAudience;
  message_template: string;
  status?: CampaignStatus;
  priority?: CampaignPriority;
  schedule?: string;
  triggers?: Omit<CampaignTrigger, 'id' | 'campaign_id' | 'created_at'>[];
}

export interface UpdateCampaignRequest {
  name?: string;
  description?: string;
  target_audience?: TargetAudience;
  message_template?: string;
  status?: CampaignStatus;
  priority?: CampaignPriority;
  schedule?: string;
}

export interface ExecuteCampaignRequest {
  dry_run?: boolean;
}

export interface ExecuteCampaignResponse {
  campaign_id: string;
  execution_id: string;
  status: string;
  total_targeted: number;
  messages_sent: number;
  messages_failed: number;
  execution_time_seconds: number;
  timestamp: string;
  dry_run?: boolean;
}

export interface CampaignMetricsResponse {
  campaign_id: string;
  metrics: CampaignMetric[];
  total_messages_sent: number;
  total_users_targeted: number;
  total_users_engaged: number;
  average_success_rate?: number;
}

export interface CampaignFilters {
  status?: CampaignStatus;
  priority?: CampaignPriority;
  target_audience?: TargetAudience;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface MessageTemplate {
  template: string;
  variables: string[];
  preview: string;
}

export const TEMPLATE_VARIABLES = [
  { name: 'user_name', description: "Student's name" },
  { name: 'sentiment_score', description: 'Current sentiment percentage' },
  { name: 'risk_score', description: 'Latest risk assessment score' },
  { name: 'case_count', description: 'Number of active cases' },
  { name: 'days_inactive', description: 'Days since last activity' },
] as const;

export const CAMPAIGN_STATUS_LABELS: Record<CampaignStatus, string> = {
  draft: 'Draft',
  active: 'Active',
  paused: 'Paused',
  completed: 'Completed',
  cancelled: 'Cancelled',
};

export const CAMPAIGN_STATUS_COLORS: Record<CampaignStatus, string> = {
  draft: 'bg-gray-500',
  active: 'bg-green-500',
  paused: 'bg-yellow-500',
  completed: 'bg-blue-500',
  cancelled: 'bg-red-500',
};

export const TARGET_AUDIENCE_LABELS: Record<TargetAudience, string> = {
  all_users: 'All Students',
  high_risk: 'High Risk',
  inactive: 'Inactive Users',
  recent_cases: 'Recent Cases',
  custom: 'Custom Segment',
};

export const TRIGGER_TYPE_LABELS: Record<TriggerType, string> = {
  sentiment_threshold: 'Sentiment Threshold',
  case_count: 'Case Count',
  sla_breach: 'SLA Breach',
  risk_score: 'Risk Score',
  inactivity: 'Inactivity',
};

export const OPERATOR_LABELS: Record<ConditionOperator, string> = {
  equals: 'Equals',
  not_equals: 'Not Equals',
  greater_than: 'Greater Than',
  greater_than_or_equal: 'Greater Than or Equal',
  less_than: 'Less Than',
  less_than_or_equal: 'Less Than or Equal',
};

/**
 * Campaign Execution History
 * Audit trail of campaign executions
 */
export interface CampaignExecutionHistory {
  id: string;
  campaign_id: string;
  campaign_name: string;
  executed_at: string;
  executed_by: number | null;
  total_targeted: number;
  messages_sent: number;
  messages_failed: number;
  execution_time_seconds: number;
  dry_run: boolean;
  targeted_user_ids: number[] | null;
  message_content: string;
  error_message: string | null;
}

export interface CampaignExecutionHistoryListResponse {
  items: CampaignExecutionHistory[];
  total: number;
  skip: number;
  limit: number;
}
